lib/ansible/modules/cloud/amazon/ec2_ami.py

```diff
@@ -542,69 +535,102 @@ def deregister_image(module, connection):
-                                 **get_ami_info(get_image_by_id(module, connection, image_id)))
-            module.exit_json(msg="AMI launch permissions updated.", launch_permissions=launch_permissions, set_perms=set_permissions, changed=True,
-                             **get_ami_info(get_image_by_id(module, connection, image_id)))
-        else:
-            module.exit_json(msg="AMI not updated.", launch_permissions=set_permissions, changed=False,
-                             **get_ami_info(get_image_by_id(module, connection, image_id)))
-    except botocore.exceptions.ClientError as e:
-        module.fail_json(msg="Error updating image - " + str(e), exception=traceback.format_exc(), **camel_dict_to_snake_dict(e.response))
+        to_add = [dict(Group=group) for group in to_add_groups] + [dict(UserId=user_id) for user_id in to_add_users]
+        to_remove = [dict(Group=group) for group in to_remove_groups] + [dict(UserId=user_id) for user_id in to_remove_users]
+
+        if to_add or to_remove:
+            try:
+                connection.modify_image_attribute(ImageId=image_id, Attribute='launchPermission',
+                                                  LaunchPermission=dict(Add=to_add, Remove=to_remove))
+                changed = True
+            except (botocore.exceptions.BotoCoreError, botocore.exceptions.ClientError) as e:
+                module.fail_json_aws(e, msg="Error updating launch permissions")
+
+    desired_tags = module.params.get('tags')
+    current_tags = boto3_tag_list_to_ansible_dict(image.get('Tags'))
+    tags_to_add, tags_to_remove = compare_aws_tags(current_tags, desired_tags, purge_tags=module.params.get('purge_tags'))
+
+    if tags_to_remove:
+        try:
+            connection.delete_tags(Resources=[image_id], Tags=[dict(Key=tagkey) for tagkey in tags_to_remove])
+            changed = True
+        except (botocore.exceptions.BotoCoreError, botocore.exceptions.ClientError) as e:
+            module.fail_json_aws(e, msg="Error updating tags")
+
+    if tags_to_add:
+        try:
+            connection.create_tags(Resources=[image_id], Tags=ansible_dict_to_boto3_tag_list(tags_to_add))
+            changed = True
+        except (botocore.exceptions.BotoCoreError, botocore.exceptions.ClientError) as e:
+            module.fail_json_aws(e, msg="Error updating tags")
+
+    description = module.params.get('description')
+    if description and description != image['Description']:
+        try:
+            connection.modify_image_attribute(Attribute='Description ', ImageId=image_id, Description=dict(Value=description))
+            changed = True
+        except (botocore.exceptions.BotoCoreError, botocore.exceptions.ClientError) as e:
+            module.fail_json_aws(e, msg="Error setting description for image %s" % image_id)
+
+    if changed:
+        module.exit_json(msg="AMI launch permissions updated.", changed=True,
```

Comment:
Is this msg necessary for backwards compatibility? Seems a little misleading since launch permissions may not have been updated.

Reply:
I don't really like the `msg` return values, but removing them seems a step too far - I've made this one more generic though